### PR TITLE
Feat 1116 location state region

### PR DESCRIPTION
**Pull request by victordelval**
## Description

Closes: regen-network/regen-web#1116

* Add `iso-3166-2` package and `@types/iso-3166-2` to populate both state/region and country selectors options
* Change Credit Retire Form jurisdiction approach (avoid using Mapbox final check with getISOString())
* Util function to get the jurisdiction iso code depending on inputs (country, region, postal code)
* Setup Jest testing in web-components project and add unit test for the jurisdiction util function

https://deploy-preview-1136--regen-registry.netlify.app/

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] provided a link to the relevant issue or specification
- [ ] provided instructions on how to test
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed
- [ ] once the PR is closed, set up backport PRs for `redwood` and `hambach` (see below)

#### Setting up backport PRs

After merging your PR to `master`, set up backports by doing the following:

1. If your branch does not have merge commits, add the following comment to
   your PR, `@Mergifyio backport redwood hambach`.

2. If your branch does have merge commits:

    a. Pull latest `master`, `hambach` and `redwood` branches

    b. Create new branches for backports and merge `master` (replace `<PR#>` with your PR #)
    ```
    git checkout -b hambach-backport-<PR#> hambach
    git merge master
    git push origin hambach-backport-<PR#>
    git checkout -b redwood-backport-<PR#> redwood
    git merge master
    git push origin redwood-backport-<PR#>`
    ```

    c. Open new PRs in regen-web targeting `hambach` and `redwood`, respectively.
  

### How to test

1.

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: web-components/src/components/inputs/LocationStateField.tsx [hunk 1]
```diff
@@ -22,53 +22,41 @@ const LocationStateField: React.FC<FieldProps> = ({
...
     }
 
-    searchState(country).then(options => {
-      if (!options || !Array.isArray(options)) return;
+    const countrySubdivisions = iso3166.country(country);
+
+    const options = Object.keys(countrySubdivisions?.sub || {})
+      .map(isoCode => ({
+        value: isoCode,
+        label: `${countrySubdivisions?.sub[isoCode].name} (${isoCode}, ${countrySubdivisions?.sub[isoCode].type})`,
```
Comment (mhagel) [inline web-components/src/components/inputs/LocationStateField.tsx:37]: Probably just need the name, but this is useful for testing purposes.

**Review by mhagel — APPROVED** [review]
I've been testing, and this is working well!

File: web-registry/src/pages/Marketplace/Storefront/hooks/useBuySellOrderSubmit.tsx [hunk 2]
```diff
@@ -68,7 +67,11 @@ const useBuySellOrderSubmit = ({
             quantity: String(creditCount),
             retirementJurisdiction: disableAutoRetire
               ? ''
-              : `${country}${stateProvinceValue}${postalCodeValue}`,
+              : getJurisdictionIsoCode({
```
Comment (mhagel) [inline web-registry/src/pages/Marketplace/Storefront/hooks/useBuySellOrderSubmit.tsx:70]: 👍

**Comment by victordelval** [comment]
Hi @erikalogie 

now in the state/region dropdown we are showing all the official subdivisions (iso-3166-2), in principle for all the countries, according to the UN / iso-3166

Some countries, like France, have more than 100 options, or even on rare occasions options with the same name, but different type of subdivision and therefore iso code...

Do we show the code and type of subdivision in parentheses, accompanying the name of the region?

state_region_selector_options

File: web-components/src/utils/locationStandard.ts [hunk 3]
```diff
@@ -49,3 +48,45 @@ export const getISOString = async (
...
+
+  // check postal code exists
+  if (_postalCode) {
+    jurisdiction = `${stateProvince} ${_postalCode}`;
+  } else {
+    jurisdiction = stateProvince as string;
+  }
+
+  return jurisdiction;
+};
```
Comment (haveanicedavid) [inline web-components/src/utils/locationStandard.ts:92]: variable reassignment here makes the function a bit hard to follow, becuase you have to understand the entire scope to know what the return value will be.  If you check the same conditions and return once they aren't met, it’s easier to follow the logic:

```suggestion
export const getJurisdictionIsoCode = ({
  country, // US
  stateProvince, // US-CO
  postalCode,
}: LocationType): string => {
  if (!iso3166.country(country)) throw new Error(`Invalid country: ${country}`);
  if (!stateProvince || iso3166.subdivision(stateProvince)) return country;
  if (!postalCode) return stateProvince;
  return `${stateProvince} ${postalCode.trim()}`;
};
```
Also not sure we want this to return an empty string, so replaced with an `Error`

Comment (victordelval) [reply web-components/src/utils/locationStandard.ts:92]: That seems very good to me. Thanks for the proposed improvements @haveanicedavid !

**Comment by victordelval** [comment]
I have configured jest in the web-components project (we also have a global command to do `yarn test-components`) and added a unit test suite for the `getJurisdictionIsoCode` util function in the `locationStandard` module-file

**Comment by erikalogie** [comment]
Is it easy to just show the type of subdivision and not the code? The code I don't think means much to most people. If that is too difficult, let's just leave this as-is and we can improve later.

**Comment by victordelval** [comment]
@erikalogie something like this?

location_region_input

**Comment by erikalogie** [comment]
@victordelval yes maybe that is ok to solve the issues with France and Spain that you saw

File: web-components/src/components/inputs/LocationCountryField.tsx [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import React, { useEffect, useState } from 'react';
 import { Field } from 'formik';
+import iso3166 from 'iso-3166-2';
```
Comment (flagrede) [inline web-components/src/components/inputs/LocationCountryField.tsx:3]: Should we lazy load this component now, since this new lib is a bit heavy and is used only in inputs?

The idea is just to replace `LocationCountryField.tsx` import by `const LocationCountryField = lazy(() => import('./LocationCountryField'));` where it's used and wrap it with `<Suspense><LocationCountryField /></Suspense>`.

File: web-components/src/components/inputs/LocationStateField.tsx [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import React, { useEffect, useState } from 'react';
-import axios from 'axios';
 import { Field, useFormikContext } from 'formik';
+import iso3166 from 'iso-3166-2';
```
Comment (flagrede) [inline web-components/src/components/inputs/LocationStateField.tsx:3]: same here

File: web-components/src/utils/locationStandard.test.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import { getJurisdictionIsoCode } from './locationStandard';
+
+describe('Location Standard utilities module', () => {
```
Comment (flagrede) [inline web-components/src/utils/locationStandard.test.ts:3]: Nice coverage 👍

**Review by flagrede — APPROVED** [review]
Nice work on this one, was not an easy one!
Just one comment, pre-approving.

File: web-components/src/utils/locationStandard.test.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import { getJurisdictionIsoCode } from './locationStandard';
```
Comment (mhagel) [inline web-components/src/utils/locationStandard.test.ts:1]: We can consider making `.spec` our convention - that is what I see most commonly. But `.test` is fine too.

Comment (mhagel) [reply web-components/src/utils/locationStandard.test.ts:1]: Thanks so much for setting up tests!

**Comment by victordelval** [comment]
@Mergifyio backport redwood hambach

File: web-components/src/components/inputs/LocationCountryField.tsx (hunk 4 above)
Comment (victordelval) [reply web-components/src/components/inputs/LocationCountryField.tsx:3]: Thanks, good advice! I accidentally merged to close the PR even though I had previously seen the review comments. They come in another little review PR

File: web-components/src/utils/locationStandard.test.ts (hunk 7 above)
Comment (victordelval) [reply web-components/src/utils/locationStandard.test.ts:1]: I don't have a strong preference, both are convention. Currently the project is using `.test.` , but they are only a few cases. Let's talk about it at the next team meeting.

File: web-components/src/components/inputs/LocationCountryField.tsx (hunk 4 above)
Comment (victordelval) [reply web-components/src/components/inputs/LocationCountryField.tsx:3]: Review PR: https://github.com/regen-network/regen-web/pull/1163
